charms/candid/src/charm.py

```diff
@@ -0,0 +1,323 @@
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/docs/sdk
+
+"""Charm the service.
+
+Refer to the following post for a quick-start guide that will help you
+develop a new k8s charm using the Operator Framework:
+
+    https://discourse.charmhub.io/t/4208
+"""
+
+import logging
+import subprocess
+from collections.abc import MutableMapping
+
+import pgsql
+from charms.operator_libs_linux.v1.snap import Snap, SnapCache, SnapError
+from ops.charm import CharmBase, StartEvent
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    WaitingStatus,
+)
+
+logger = logging.getLogger(__name__)
+
+SNAP_NAME = "candid"
+
+REQUIRED_SETTINGS = [
+    "admin-agent-public-key",
+    "api-macaroon-timeout",
+    "discharge-macaroon-timeout",
+    "discharge-token-timeout",
+    "identity-providers",
+    "location",
+    "private-key",
+    "public-key",
+    "rendezvous-timeout",
+]
+
+
+class CandidCharm(CharmBase):
+    """Charm the service."""
+
+    _stored = StoredState()
+
+    @property
+    def snap(self) -> Snap:
+        """Retrieves snap from the snap cache"""
+        return SnapCache().get(SNAP_NAME)
+
+    @property
+    def snap_installed(self) -> bool:
+        """Reports if the Candid snap is installed."""
+        return self.snap.present
+
+    @property
+    def snap_running(self):
+        """Reports if the 'candidsrv' snap daemon is running."""
+        return self.snap.services["candidsrv"]["active"]
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        # Hooks
+        self.framework.observe(self.on.install, self._install)
+        self.framework.observe(self.on.start, self._start)
+        self.framework.observe(self.on.config_changed, self._config_changed)
+        self.framework.observe(
+            self.on.website_relation_joined, self._on_website_relation_joined
+        )
+
+        # Database
+        self.db = pgsql.PostgreSQLClient(self, "postgres")
+        self.framework.observe(
+            self.db.on.database_relation_joined,
+            self._on_database_relation_joined,
+        )
+        self.framework.observe(
+            self.db.on.master_changed, self._on_master_changed
+        )
+
+        self._stored.set_default(db_uri=None)
+
+    ###################
+    # LIFECYCLE HOOKS #
+    ###################
+    def _install(self, _):
+        """
+        Install candid snap
+        """
+        self.set_status_and_log("Installing snap", WaitingStatus)
+        self._install_snap()
+        if self.snap_installed:
+            self.set_status_and_log("Snap installed", WaitingStatus)
+
+    def _start(self, event: StartEvent):
+        """
+        Starts candidsrv service
+        """
+        if self._check_config(event):
+            self.set_status_and_log("Starting candid", WaitingStatus)
+            self.snap.start(["candidsrv"])
```

Comment:
Based on this - https://juju.is/docs/sdk/start-event
"The start event is triggered on each starting unit immediately after the first config-changed event"
But we already restart/start the service in the _config_changed handler. So it might be a bit redunant what we're doing here. I would imagine just running the check
```
if self.snap_running:
                self.set_status_and_log("Ready ", ActiveStatus)
```
is sufficient.